Sources/NIOCore/EventLoopFuture+AssumeIsolated.swift

```diff
@@ -0,0 +1,510 @@
+            return wrapped.always {
+                unsafeTransfer.wrappedValue($0)
+            }
+        }
+
+        /// Unwrap an `EventLoopFuture` where its type parameter is an `Optional`.
+        ///
+        /// Unwraps a future returning a new `EventLoopFuture` with either: the value passed in the `orReplace`
+        /// parameter when the future resolved with value Optional.none, or the same value otherwise. For example:
+        /// ```
+        /// promise.futureResult.unwrap(orReplace: 42).wait()
+        /// ```
+        ///
+        /// - parameters:
+        ///     - orReplace: the value of the returned `EventLoopFuture` when then resolved future's value is `Optional.some()`.
+        /// - returns: an new `EventLoopFuture` with new type parameter `NewValue` and the value passed in the `orReplace` parameter.
+        @inlinable
+        func unwrap<NewValue>(
+            orReplace replacement: NewValue
+        ) -> EventLoopFuture<NewValue> where Value == Optional<NewValue> {
+            return self.map { (value) -> NewValue in
+                guard let value = value else {
+                    return replacement
+                }
+                return value
+            }
+        }
+
+        /// Unwrap an `EventLoopFuture` where its type parameter is an `Optional`.
+        ///
+        /// Unwraps a future returning a new `EventLoopFuture` with either: the value returned by the closure passed in
+        /// the `orElse` parameter when the future resolved with value Optional.none, or the same value otherwise. For example:
+        /// ```
+        /// var x = 2
+        /// promise.futureResult.unwrap(orElse: { x * 2 }).wait()
+        /// ```
+        ///
+        /// - parameters:
+        ///     - orElse: a closure that returns the value of the returned `EventLoopFuture` when then resolved future's value
+        ///         is `Optional.some()`.
+        /// - returns: an new `EventLoopFuture` with new type parameter `NewValue` and with the value returned by the closure
+        ///     passed in the `orElse` parameter.
+        @inlinable
+        func unwrap<NewValue>(
+            orElse callback: @escaping () -> NewValue
+        ) -> EventLoopFuture<NewValue> where Value == Optional<NewValue> {
+            return self.map { (value) -> NewValue in
+                guard let value = value else {
+                    return callback()
+                }
+                return value
+            }
+        }
+    }
+
+    /// Assumes the calling context is isolated to the future's event loop.
+    @usableFromInline
+    func assumeIsolated() -> AssumeIsolated {
+        self.eventLoop.assertInEventLoop()
+        return AssumeIsolated(wrapped: self)
+    }
+}
+
+
+extension EventLoopPromise {
+    /// A struct wrapping an ``EventLoopPromise`` that ensures all calls to any method on this struct
+    /// are coming from the event loop of the promise.
+    @usableFromInline
+    struct AssumeIsolated {
```

Comment:
> I intentionally did not return the AssumeIsolated from the methods but rather the regular types since those types contain way more methods and I don't want to duplicate everything.

In that case why don't we have an `unwrap` API or similar which returns the non-isolated future? That makes the API symmetric when the isolation changes: i.e. you explicitly have to do so.